Lib/fontTools/otlLib/builder.py

```diff
@@ -453,45 +455,47 @@ def buildFormat2Subtable(self, ruleset, classdefs, chaining=True):
         classSets = []
         for _ in inClasses:
             classSet = self.newRuleSet_(format=2, chaining=chaining)
-            classSet.populateDefaults()
             classSets.append(classSet)
 
         coverage = set()
         classRuleAttr = self.ruleAttr_(format=2, chaining=chaining)
 
         for rule in ruleset.rules:
-            rule_table = self.newRule_(format=2, chaining=chaining)
+            ruleAsSubtable = self.newRule_(format=2, chaining=chaining)
             if chaining:
-                rule_table.BacktrackGlyphCount = len(rule.prefix)
-                rule_table.LookAheadGlyphCount = len(rule.suffix)
-                rule_table.Backtrack = [
+                ruleAsSubtable.BacktrackGlyphCount = len(rule.prefix)
+                ruleAsSubtable.LookAheadGlyphCount = len(rule.suffix)
+                # The glyphs in the rule may be list, tuple, odict_keys...
```

Comment:
all the types you mentioned keep their order, so the comment that follows about order not being important is misplaced. Only if these were unordered set or frozenset it would be relevant.
I find it strange that we are so inconsistent we collection types. Can't we make sure we store these rules using the same type, instead of casting them ex-post?

Reply:
IIUC, the reason you do `list(x)[0]` is to be able to get the first item of a collection which can sometimes be list, tuple or odict_keys.
I wonder: is there a good reason why the type used for the rule.prefix/glyphs/suffix can be either of these? I don't think there is, it's just happens to be that way. I suggest that we enforce one of these types and stick with it in a consistent way.
Do we want an immutable sequence that keeps order and that we can index into, then we should use tuple only.
I don't like this overly relaxed approach to typing, that induces one to do extra copies because anything goes.

Reply:
A more efficient way to get the first item of a collection would be to do `next(iter(x))`, so you don't have to construct a whole list just to get the first item, and would work for odict_keys or (randomly ordered) sets too. In anycase, I reiterate my preference for sticking to one type only for the same kind of data, in this case a rule's glyph names.

Reply:
I agree that we should tidy up what goes in the rules. Can I suggest, though, that we make an issue for that and address it in a separate PR? It's going to require changes to the parts of feaLib.builder and feaLib.ast which are creating these ChainContextualRules, and that's not directly related to what I'm doing here.

Reply:
sgtm

Reply:
Thanks. Are you happy with this PR on that basis, then?